로그인, 회원가입과 유저 CRUD 구현

jsham0414 opened the pull request:
## 새로운 기능
#### Auth
- 회원가입
- 로그인
#### User
- 유저 정보 수정
- 유저 정보 보기
- 회원탈퇴
#### Security
- Encrypt.java
   - 어노테이션 클래스
- EncryptData.java
   - PointCut으로 특정 메서드 실행 시 @Encrypt이 적용 된 String 타입의 필드는 암, 복호화됩니다.
- EncryptString.java
   - Base64를 사용해 String 타입을 AES 방식으로 암호화합니다.

## 중점적으로 봐주었으면 하는 부분
#### AuthController.java
- 의존성 순환 문제 때문에 여기서 토큰 생성을 합니다. 컨트롤러에서 생성하는건 좋지 않은 것 같은데 좋은 대안이 있을까요?
#### UpdateUserInfo.java
- 이 클래스도 DTO라고 볼 수 있나요? 아니라면 어떤 패키지에 넣는게 좋을지 알고 싶습니다.
#### User.java
- roles는 User 조회 직후에 AUTHORITY 테이블에서 가져오는데 성능면에서 문제가 된다면 더 좋은 대안이 있는지 궁금합니다.

jsham0414 committed:
Feat:회원가입, 로그인 추가

유저 회원가입 로그인 구현
필드 암, 복호화를 위한 Encrypt 어노테이션 구현

jsham0414 committed:
docs: README 수정

jsham0414 committed:
Feat: 유저 CRUD 추가

유저 CRUD 구현

jsham0414 committed:
Test: 테스트 코드 추가

AuthController, AuthSercvice, UserController, UserService에 대한 테스트 코드 추가

spring-kotlin-lover commented:
* AuthController.java
> 의존성 순환 문제 때문에 여기서 토큰 생성을 합니다. 컨트롤러에서 생성하는건 좋지 않은 것 같은데 좋은 대안이 있을까요?

의존성 순환 문제가 어떤 말씀이신가요? 어떤 부분에서 의존성 순환이 되는지 궁금하네요

* UpdateUserInfo.java
> 이 클래스도 DTO라고 볼 수 있나요? 아니라면 어떤 패키지에 넣는게 좋을지 알고 싶습니다.

정확히는 Controller 의 DTO 라고 부르는게 정확합니다. 지금처럼 dto 패키지에 두시거나 controller.dto 패키지로 가는 것이 좋아보입니다.

* User.java
> roles는 User 조회 직후에 AUTHORITY 테이블에서 가져오는데 성능면에서 문제가 된다면 더 좋은 대안이 있는지 궁금합니다.

테이블 join 연산이 들어가니 당연히 성능적으로 문제가 되겠죠. 이런 경우에는 roles 는 User 테이블에서 column 속성으로 들어가게 하면 됩니다.
테이블을 없애고 collection 변환기를 통해 저장하면 되죠. attribute converter 를 구현하신다면 컬렉션 변환에 대한 문제도 해결 가능할 것으로 보이네요.
https://www.baeldung.com/jpa-attribute-converters

spring-kotlin-lover commented on api/src/main/java/com/zerobase/commerce/api/exception/ErrorCode.java:
```diff
@@ -0,0 +1,30 @@
+
+@Getter
+public enum ErrorCode {
+    UNEXPECTED_ERROR(HttpStatus.BAD_REQUEST, "Unexpected error occurred."),
```
BAD_REQUEST 를 사용하는 것이 아닌 INTERNAL_SERVER_ERROR 를 사용하는 것이 맞아보입니다. 

서버에서 알 수 없는 에러로 client 에게 응답을 내려줄때는 보통 500 http status 를 사용하게 된답니다.
https://developer.mozilla.org/ko/docs/Web/HTTP/Status/500

spring-kotlin-lover commented on api/src/main/java/com/zerobase/commerce/api/exception/ErrorCode.java:
```diff
@@ -0,0 +1,30 @@
+    INVALID_USER_ID(HttpStatus.BAD_REQUEST, "The ID you entered is an ID that does not exist. Please enter a different ID."),
+    SIGN_IN_FAILED(HttpStatus.BAD_REQUEST, "Login failed. Please check your input again."),
+    PASSWORD_INCORRECT(HttpStatus.BAD_REQUEST, "Password is not equal. Please check your input again."),
+    INVALID_TOKEN(HttpStatus.BAD_REQUEST, "Invalid token. Please check again.");
```
일반적으로 token 값이 올바르지는 않을때는 http status 400 보다는 401 을 내려주는 것이 맞습니다. 
올바르지 않는 인증 token 을 요청해서 자격이 없다는 것을 나타내는 것이니까요

https://developer.mozilla.org/ko/docs/Web/HTTP/Status/401

spring-kotlin-lover commented on api/src/main/java/com/zerobase/commerce/api/user/controller/AuthController.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @PostMapping("/sign-in")
+    ResponseEntity<?> signIn(@RequestBody AuthDto.Request.SignIn request) {
```
서버 개발자는 클라이언트가 요청한 값을 그대로 활용해서는 안됩니다. 어떤 값이 요청될지도 모르고 다양한 데이터에 의해 DB 에 장애를 일으킬 수 있기 때문에 validation 을 매우 꼼꼼하게 해주셔야 되요. 그래서 spring-boot-starter-validation 모듈을 참고해서 값 검증을 할 수 있도록 진행해보시면 좋겠네요

https://velog.io/@_koiil/SpringBoot-Spring-Validation%EC%9D%84-%EC%9D%B4%EC%9A%A9%ED%95%9C-%EC%9C%A0%ED%9A%A8%EC%84%B1-%EA%B2%80%EC%A6%9D

jsham0414 replied:
@spring-kotlin-lover 모든 리뷰 확인했습니다. 감사합니다 멘토님 !!
의존성 순환 문제는 TokenProvider에서 AuthService로 Username 확인을 하고, AuthService에서 TokenProvider로 토큰 생성을 해서 생기는 문제였습니다.
TokenProvider에서 토큰 생성과 토큰 확인의 기능을 모두 수행하는 건 아닌 것 같아 TokenAuthenticator를 추가해서 해결했습니다.

spring-kotlin-lover replied:
잘 고민해주시고 개선해주셨네요 ^^ 고생많았습니다. Token 관련된 발급/인증 로직도 각 역할에 맞는 2개의 클래스로 구분해서 좋았네요

jsham0414 committed:
Feat: PR 리뷰 후속 커밋

Validation과 전역 예외 처리 구현
AuthorityStatusConverter 구현
ErrorCode 상황에 맞는 HttpStatus로 수정

jsham0414 committed:
Refactor: 의존성 순환 해결

AuthService와 TokenProvider의 의존성 순환을 해결하기 위해 TokenAuthenticator 추가

jsham0414 commented:
## 수정된 사항
- Validation
   - Controller에서 Validation 처리와 전역 예외 처리 구현
- AuthorityStatusConverter.java
   - roles를 role 테이블 join으로 가져오던 것을 AttributeConverter로 자동으로 변환
- ErrorCode.java
   - 상황에 맞는 HttpStatus로 수정
- 의존성 순환
   - AuthService와 TokenProvider의 의존성 순환을 해결하기 위해 TokenAuthenticator 추가

jsham0414 committed:
Refactor: AuthorityStatusConverter 패키지명 수정

domain에서 constant로 옮김
